### PR TITLE
Compose source maps

### LINKED ISSUE #116: Enable source maps in v8 stack traces
Looks like this will do it:

https://github.com/evanw/node-source-map-support

### PR DESCRIPTION
**Pull request by LucasSloan**
Currently, tsickle produces a source map for its transformations and tsc produces a source map for its transformations, but we don't produce a merged source map that maps the js to the original ts.  This PR adds support for merging the two source maps as well as e2e tests for the transformations.

Closes #116

### CODE REVIEW

File: test/source_map_test.ts [hunk 1]
```diff
@@ -6,10 +6,14 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
-import {expect} from 'chai';
```
Comment (LucasSloan) [inline test/source_map_test.ts]: Remove all the changes from this file.

**Comment by LucasSloan** [comment]
I signed it!

**Comment by LucasSloan** [comment]
I signed it!

File: src/main.ts [hunk 2]
```diff
@@ -163,14 +164,19 @@ function createSourceReplacingCompilerHost(
  * Compiles TypeScript code into Closure-compiler-ready JS.
  * Doesn't write any files to disk; all JS content is returned in a map.
  */
-function toClosureJS(
+export function toClosureJS(
     options: ts.CompilerOptions, fileNames: string[], settings: Settings,
-    allDiagnostics: ts.Diagnostic[]): {jsFiles: Map<string, string>, externs: string}|null {
+    allDiagnostics: ts.Diagnostic[],
```
Comment (rkirov) [inline src/main.ts]: did you run clang-format over the code, that argument packing looks a bit suspicious here.

File: src/main.ts [hunk 3]
```diff
@@ -163,14 +164,19 @@ function createSourceReplacingCompilerHost(
  * Compiles TypeScript code into Closure-compiler-ready JS.
  * Doesn't write any files to disk; all JS content is returned in a map.
```
Comment (rkirov) [inline src/main.ts]: comment on fileNames vs files function API. I find it a bit odd to be able to pass both.

File: package.json [hunk 4]
```diff
@@ -42,7 +42,7 @@
     "merge2": "^1.0.2",
     "temp": "^0.8.1",
     "tslint": "^3.15.1",
-    "typescript": "^2.0.0"
+    "typescript": "~2.0.0"
```
Comment (rkirov) [inline package.json]: is this change needed?

File: src/main.ts [hunk 5]
```diff
@@ -227,6 +235,26 @@ function toClosureJS(
       let {output} = tsickle.processES5(
           fileName, moduleId, jsFiles.get(fileName)!, cliSupport.pathToModuleName);
       jsFiles.set(fileName, output);
+    } else {
+      const tscSourceMapJson: any = jsFiles.get(fileName);
```
Comment (rkirov) [inline src/main.ts]: why not keep it :string, which is the inferred type.

File: src/main.ts [hunk 6]
```diff
@@ -227,6 +235,26 @@ function toClosureJS(
...
+      const tscSourceMap = new SourceMapConsumer(tscSourceMapJson);
+      const tscSourceMapGenerator = SourceMapGenerator.fromSourceMap(tscSourceMap);
+
+      for (const sourceFileName of (new SourceMapConsumer(tscSourceMapJson) as any).sources) {
```
Comment (rkirov) [inline src/main.ts]: why as any here?

File: test/e2e_source_map_test.ts [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+      let x : string = 'a string';
+      let y : string = 'another string';
+      let z : string = x + y;`);
+
+    // Run tsickle+TSC to convert inputs to Closure JS files.
+    const closure = toClosureJS(
+        {sourceMap: true} as ts.CompilerOptions, ['input.ts'], {isUntyped: false} as Settings,
+        diagnostics, sources);
+
+    if (!closure) {
```
Comment (rkirov) [inline test/e2e_source_map_test.ts]: extract the repeated bits (asserting !closure and !compiledJS) into a helper method.

File: test/e2e_source_map_test.ts [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    if (!compiledJS) {
+      assert.fail();
+      return;
+    }
+
+    const lines = compiledJS.split('\n');
+    const stringXLine = lines.findIndex(l => l.indexOf('a string') !== -1) + 1;
+    const stringXColumn = lines[stringXLine - 1].indexOf('a string') + 1;
+    const stringBLine = lines.findIndex(l => l.indexOf('fourth rate') !== -1) + 1;
+    ;
```
Comment (rkirov) [inline test/e2e_source_map_test.ts]: extra ;

File: test/e2e_source_map_test.ts [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    expect(sourceMap.originalPositionFor({line: stringYLine, column: stringYColumn}).source)
+        .to.equal('input.ts', 'input file name');
+  });
+
+  it('composes sources maps with multiple input files', function() {
+    const diagnostics: ts.Diagnostic[] = [];
+
+    const sources = new Map<string, string>();
+    sources.set(ts.sys.resolvePath('input1.ts'), `
+      class X { field: number; }
```
Comment (rkirov) [inline test/e2e_source_map_test.ts]: we do some funky stuff with decorators (see tsickle/test_files/decorator/decorator.ts), can you throw in a test for one of those. If it requires more fixing feel free to handle in a next PR.

File: src/main.ts (hunk 2 above)
Comment (LucasSloan) [inline src/main.ts]: I did.

File: src/main.ts (hunk 5 above)
Comment (LucasSloan) [inline src/main.ts]: new SourceMapConsumer() takes in either a string or an object, but the d.ts only has the object version.

File: src/main.ts (hunk 6 above)
Comment (LucasSloan) [inline src/main.ts]: Again, the d.ts doesn't cover the fact that the SourceMapConsumer object has a sources field.

**Comment by LucasSloan** [comment]
I signed it!

**Comment by LucasSloan** [comment]
I signed it!

File: src/main.ts (hunk 3 above)
Comment (LucasSloan) [inline src/main.ts]: done.

File: test/e2e_source_map_test.ts (hunk 7 above)
Comment (LucasSloan) [inline test/e2e_source_map_test.ts]: done.

File: test/e2e_source_map_test.ts (hunk 8 above)
Comment (LucasSloan) [inline test/e2e_source_map_test.ts]: done.

File: test/e2e_source_map_test.ts (hunk 7 above)
Comment (LucasSloan) [inline test/e2e_source_map_test.ts]: done.

**Comment by LucasSloan** [comment]
I signed it!

**Comment by evmar** [comment]
Meta-comment: we have been using the "squash and merge" git workflow, which means all of your commits will become a single one.  If you want to land some of the other changes separately (like pinning the version, which LGTM) you could maybe cherry-pick them into a separate pull request.

File: src/main.ts (hunk 2 above)
Comment (evmar) [inline src/main.ts]: The test suite ensures everything is clang-formatted. :)

File: src/main.ts (hunk 5 above)
Comment (evmar) [inline src/main.ts]: Can we fix the d.ts?  (Not necessary for this change, but it would be nice.)

File: src/main.ts [hunk 10]
```diff
@@ -227,6 +237,26 @@ function toClosureJS(
...
+        const tsickleSourceMapGenerator = tsickleSourceMaps.get(sourceFileName);
+        if (!tsickleSourceMapGenerator) {
+          console.log('couldn\'t find a souce map for: ' + sourceFileName);
```
Comment (evmar) [inline src/main.ts]: console.warn maybe?

File: src/main.ts (hunk 5 above)
Comment (LucasSloan) [inline src/main.ts]: I'm planning to.

**Comment by LucasSloan** [comment]
I'll give that some thought.  Could you point me at the internal runner?

File: test/e2e_source_map_test.ts (hunk 9 above)
Comment (LucasSloan) [inline test/e2e_source_map_test.ts]: What did you mean by this?  I'm having trouble figuring out what it would mean to test decorations.

Comment (LucasSloan) [inline test/e2e_source_map_test.ts]: Done.

File: src/main.ts (hunk 10 above)
Comment (mprobst) [inline src/main.ts]: nit: prefer template strings to avoid the escaping

File: test/e2e_source_map_test.ts [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    class DecoratorTest {
+      public x(s: string): string { return s; }
+    }`);
+
+    const {compiledJS, sourceMap} = compile(sources);
+
+    const xPosition = getLineAndColumn(compiledJS, 'x');
+
+    expect(sourceMap.originalPositionFor(xPosition).line).to.equal(6, 'method X position');
+
```
Comment (mprobst) [inline test/e2e_source_map_test.ts]: nit: spurious empty line?

File: test/e2e_source_map_test.ts [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    expect(sourceMap.originalPositionFor({line: stringBLine, column: stringBColumn}).line)
+        .to.equal(4, 'fouth string definition');
+    expect(sourceMap.originalPositionFor({line: stringBLine, column: stringBColumn}).source)
+        .to.equal('input2.ts', 'second input file');
+  });
+
+  it('handles decorators correctly', function() {
+    const sources = new Map<string, string>();
+    sources.set('input.ts', `/** @Annotation */
+    function classAnnotation(t: any) { return t; }
```
Comment (mprobst) [inline test/e2e_source_map_test.ts]: can you indent these literals a bit from the regular code?

File: test/e2e_source_map_test.ts [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+    expect(sourceMap.originalPositionFor({line: stringYLine, column: stringYColumn}).line)
+        .to.equal(4, 'second string definition');
+    expect(sourceMap.originalPositionFor({line: stringYLine, column: stringYColumn}).source)
+        .to.equal('input.ts', 'input file name');
+  });
+
+  it('composes sources maps with multiple input files', function() {
+    const sources = new Map<string, string>();
+    sources.set('input1.ts', `
+      class X { field: number; }
```
Comment (mprobst) [inline test/e2e_source_map_test.ts]: please indent +4 from the left margin, i.e. two more spaces here and below

File: src/main.ts [hunk 14]
```diff
@@ -227,6 +237,26 @@ function toClosureJS(
...
+          return null;
+        }
+        const tsickleRawSourceMap = tsickleSourceMapGenerator.toJSON();
+        tsickleRawSourceMap.sources[0] = sourceFileName;
```
Comment (mprobst) [inline src/main.ts]: Add a comment on why you need to update the file name here?

File: test/e2e_source_map_test.ts [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+    const xPosition = getLineAndColumn(compiledJS, 'x');
+
+    expect(sourceMap.originalPositionFor(xPosition).line).to.equal(6, 'method X position');
+
+  });
+});
+
+function getLineAndColumn(source: string, token: string): {line: number, column: number} {
+  const lines = source.split('\n');
+  const line = lines.findIndex(l => l.indexOf(token) !== -1) + 1;
```
Comment (mprobst) [inline test/e2e_source_map_test.ts]: throw when you cannot find a line that contains `token`?

**Comment by mprobst** [comment]
Regarding the code layering. I agree with @evmar, we should keep functionality below `main.ts`.

Currently, plain `tsc` has a bunch of options to control source map creation (not at all, separate files, embed in generated `.js` sources). Maybe the most elegant interface for the user would be to transparently wrap this?

That is, if `tsconfig.json` asks for embedding source maps, just embed the correct source map, and when `tsconfig.json` asks for writing source maps to disk, write the right source maps?

File: test/e2e_source_map_test.ts (hunk 15 above)
Comment (LucasSloan) [inline test/e2e_source_map_test.ts]: Done.

File: src/main.ts (hunk 10 above)
Comment (LucasSloan) [inline src/main.ts]: Done.

File: src/main.ts (hunk 14 above)
Comment (LucasSloan) [inline src/main.ts]: That was super confusing, I ended up changing rewriter.ts to make it unnecessary.
